lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp

```diff
@@ -315,6 +380,150 @@ struct MMAV3UseRegOperand
+//  elementwise(convert(x, #dot_operand)).
+//
+// Whereas (MMAV2) HoistLayoutConversion hoists thru one op at a time and requires
+// multiple passes, this will directly hoist the convert to the right place in one pass.
+struct MMAV3HoistLayoutConversion
+    : public OpRewritePattern<triton::nvidia_gpu::WarpGroupDotOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(triton::nvidia_gpu::WarpGroupDotOp dotOp,
+                                PatternRewriter &rewriter) const override {
+    // Can only hoist operand 0
+    auto alloc = dotOp.getOperand(0).getDefiningOp<LocalAllocOp>();
+    if (!alloc || !alloc.getSrc())
+      return failure();
+
+    auto getEncoding = [](Value v) {
+      return cast<TensorOrMemDesc>(v.getType()).getEncoding();
+    };
+
+    if (!isa<SharedEncodingAttr>(getEncoding(dotOp.getOperand(0))))
+      return failure();
+
+    // Step 1: Performs checks for early stop
+    Type inputEltTy;
+    auto srcEnc = dyn_cast<BlockedEncodingAttr>(getEncoding(alloc.getSrc()));
+    auto dstEnc =
+        dyn_cast<NvidiaMmaEncodingAttr>(getEncoding(dotOp.getResult()));
+    // Want: A's Encoding to be Blocked and D's encoding to be NvidiaMmaV3
+    if (!srcEnc || !dstEnc || dstEnc.getVersionMajor() != 3)
+      return failure();
+    auto src = alloc.getSrc().getDefiningOp();
+    // Value passed to alloc must have Tensor arguments and single Tensor result
+    if (!src || src->getNumOperands() == 0 || src->getNumResults() != 1)
+      return failure();
+    if (!all_of(src->getOperandTypes(),
+                [](Type ty) { return isa<RankedTensorType>(ty); }))
+      return failure();
+    auto srcTy = dyn_cast<RankedTensorType>(src->getResult(0).getType());
+    if (!srcTy)
+      return failure();
+    inputEltTy = srcTy.getElementType();
+
+    // Check src itself can be hoisted through
+    if (!canHoistDotOpEncV3(src))
+      return failure();
+
+    // Step 2: Obtain slice of ops between load/constant and local_alloc
+    SetVector<Operation *> slice;
+    BackwardSliceOptions opt;
+    opt.omitBlockArguments = true;
+    opt.filter = [&](Operation *op) {
+      // Stop before Load, ConstantOp, or LocalLoad (which is unlikely)
+      return (op->getParentRegion() == alloc->getParentRegion())
+        && !isa<LoadOp, arith::ConstantOp, LocalLoadOp>(op)
+        && (op->getNumOperands() != 0);
+    };
+    getBackwardSlice(alloc.getOperation(), &slice, opt);
+
+    // Step 3: Verify slice can be hoisted through
+    if (slice.empty())
+      return failure();
+
+    auto isBlockedRankedTensor = [&](auto val) {
+      return isa<BlockedEncodingAttr, DotOperandEncodingAttr>(getEncoding(val)) &&
+        isa<RankedTensorType>(val.getType());
+    };
+
+    for (Operation *currOp : slice) {
+      if (!canHoistDotOpEncV3(currOp))
+        return failure();
+
+      // We previously ensured that all ops in slice have at least one operand
+      for (auto operand : currOp->getOperands()) {
```

Comment:
It's not easy for me to follow what this block of code checks. Please add a comment (and probably introduce the frontierOp term here instead of below).